src/librustc_typeck/check/demand.rs

```diff
@@ -539,6 +541,41 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                     return Some((sp, "consider removing the borrow", code));
                 }
             }
+            (
+                _,
+                &ty::RawPtr(TypeAndMut { ty: _, mutbl: hir::Mutability::Not }),
+                &ty::Ref(_, _, hir::Mutability::Not),
+            ) => {
+                let cause = self.cause(rustc_span::DUMMY_SP, ObligationCauseCode::ExprAssignable);
+                // We don't ever need two-phase here since we throw out the result of the coercion
+                let coerce = Coerce::new(self, cause, AllowTwoPhase::No);
+
+                if let Some(steps) =
+                    coerce.autoderef(sp, checked_ty).skip(1).find_map(|(referent_ty, steps)| {
+                        Some(steps).filter(|_| {
+                            coerce
+                                .unify(
+                                    coerce.tcx.mk_ptr(ty::TypeAndMut {
+                                        mutbl: hir::Mutability::Not,
+                                        ty: referent_ty,
+                                    }),
+                                    expected,
+                                )
+                                .is_ok()
+                        })
```

Comment:
This took me a moment to untangle. The filter is always executed, so I think it would be better formulated as

```suggestion
                            coerce
                                .unify(
                                    coerce.tcx.mk_ptr(ty::TypeAndMut {
                                        mutbl: hir::Mutability::Not,
                                        ty: referent_ty,
                                    }),
                                    expected,
                                )
                                .ok().map(|_| steps)
```

Reply:
Nice! I didn't expect to write like that.